src/bsk201910/rollbacks.md

```diff
@@ -0,0 +1,407 @@
+
+_There's also something else that looks useful but doesn't work quite right._
+
+The history command displays two columns: _Revision_ and _Change-Cause_.
+
+```terminal|command=1|title=bash
+kubectl rollout history deployment/app
+REVISION  CHANGE-CAUSE
+1         <none>
+2         <none>
+3         <none>
+```
+
+While you're now familiar with the Revision column, you might be wondering what Change-Cause is used for — and why it's always set to `<none>`.
+
+When you create a resource in Kubernetes, you can append the `--record` flag like so:
+
+```terminal|command=1|title=bash
+kubectl create -f deployment.yaml --record
+```
+
+When you do, Kubernetes adds an annotation to the resource with the command that generated it.
+
+In the example above, the Deployment has the following metadata section:
+
+```yaml|highlight=5|title=deployment.yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example-deployment
+  kubernetes.io/change-cause: kubectl create --filename=deployment.yaml
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      name: app
+  template:
+    metadata:
+      labels:
+        name: app
+    spec:
+      containers:
+      - name: app
+        image: learnk8s/hello:1.0.0
+```
+
+Now, if you try to display the history again, you might notice that the same annotation is used in the rollout history command:
+
+```terminal|command=1|title=bash
+kubectl rollout history app
+REVISION  CHANGE-CAUSE
+1         kubectl create --filename=deployment.yaml --record=true
+```
+
+If you change the container image in the YAML file and apply the new configuration with:
+
+```terminal|command=1|title=bash
+kubectl apply -f deployment.yaml --record
+```
+
+You should see the following new entry in the rollout history:
+
+```terminal|command=1|title=bash
+kubectl rollout history deployment/app
+REVISION  CHANGE-CAUSE
+1         kubectl create --filename=deployment.yaml --record=true
+2         kubectl apply --filename=deployment.yaml --record=true
+```
+
+The `--record` command can be used with any resource type, but it has an effect only on Deployment, DaemonSet, and StatefulSet resources, i.e. resources that can be "rolled out" (see `kubectl rollout -h`).
+
+But you should remember:
+
+- The `--record` flag adds an annotation to the YAML resource, which can be changed at any time
+- The rollout history command uses the value of this annotation to populate the Change-Cause table
+- The annotation contains the last command only. If you create the resource and later use `kubectl scale --replicas=10 deploy/app --record` to scale it, only the scaling command is stored in the annotation.
+
+Also, there is an [ongoing discussion on deprecating the `--record` flag](https://github.com/kubernetes/kubernetes/issues/40422).
+
+The feature provides little value for manual usage, but it still has some justification for automated processes as a simple form of auditing (keeping track of which commands caused which changes to a rollout).
```

Comment:
A conclusion chapter is missing, where I would mention a couple of things.

I would have a sentence to emphasize how much we have learnt and how complicated it is (e.g. holes in rollout history revisions, deployment hashes, and what not).

I think you should also have a short as possible reference to how we supposed to work: git repo for all the sysadmins and just `kubectl apply -f .` whole directories.  And then discuss that mixing the two is a bad idea, and instead you should roll back in got and always roll forward with `kubectl apply -f`.